### PR TITLE
Add a context to store enums from server

### PR DESCRIPTION
**Pull request by frozenhelium**

### CODE REVIEW

File: src/views/PerWorkPlanForm/CustomComponentInput/index.tsx [hunk 1]
```diff
@@ -103,6 +126,7 @@ function CustomComponentInput(props: Props) {
             />
             <div>
                 <Button
+                    // FIXME: add title attribute
```
Comment (tnagorra) [inline src/views/PerWorkPlanForm/CustomComponentInput/index.tsx]: We don't need this comment

File: src/views/Search/EmergencyTable/index.tsx [hunk 2]
```diff
@@ -133,12 +113,18 @@ function EmergencyTable(props: Props) {
...
+        createCountryListColumn<EmergencyResult, number>(
             'countries',
             strings.searchEmergencyTableCountry,
-            countryListColumnParams,
+            (item) => item.countries_id.map(
+                (countryId, index) => ({
+                    id: countryId,
+                    name: item.countries[index],
+                }),
+            ),
```
Comment (tnagorra) [inline src/views/Search/EmergencyTable/index.tsx]: We need to improve this.

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/views/Emergencies/EmergenciesTable/index.tsx [hunk 3]
```diff
@@ -25,24 +25,24 @@ import i18n from './i18n.json';
...
 thirtyDaysAgo.setHours(0, 0, 0, 0);
 
+const PAGE_SIZE = 5;
 const keySelector = (item: EventListItem) => item.id;
 
 function EventItemsTable() {
     const strings = useTranslation(i18n);
     const sortState = useSortState({ name: 'created_at', direction: 'dsc' });
     const { sorting } = sortState;
+    const [page, setPage] = useState(0);
```
Comment (tnagorra) [inline src/views/Emergencies/EmergenciesTable/index.tsx]: Let's confirm this is zero

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/components/AppealsTable/index.tsx [hunk 4]
```diff
@@ -220,7 +242,7 @@ function AppealsTable() {
             <SortContext.Provider value={sortState}>
                 <Table
                     pending={appealsPending}
-                    filtered={!!(displacementType && appealType)}
+                    filtered={!!(filterDisplacement && filterAppeal)}
```
Comment (tnagorra) [inline src/components/AppealsTable/index.tsx]:
```suggestion
                    filtered={!!(filterDisplacement || filterAppeal)}
```

Comment (tnagorra) [reply src/components/AppealsTable/index.tsx]: Also let's use isDefined

**Comment by tnagorra** [comment]
@frozenhelium @thenav56 

Using server `https://per-ifrc.dev.datafriendlyspace.org/` there are still 25 typeerrors.

The errors are not just on Register and Dref pages.
```
src/components/ActiveOperationMap/index.tsx:225:39 - error TS2352: Conversion of type 'string' to type '{ type: "Point"; coordinates: [number, number]; }' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
src/views/DrefApplicationForm/Operation/index.tsx:331:25 - error TS2322: Type 'string | undefined' is not assignable to type 'number | null | undefined'.
src/views/DrefApplicationForm/Operation/index.tsx:332:25 - error TS2322: Type '(...entries: EntriesAsList<{ title?: string | undefined; id?: number | undefined; status?: 0 | 1 | undefined; national_society?: number | undefined; ifrc?: string | undefined; translation_module_original_language?: string | undefined; ... 133 more ...; anticipatory_actions?: string | undefined; } & Pick<...>>) => void' is not assignable to type '(value: number | undefined, name: "disability_people_per", e?: FormEvent<HTMLInputElement> | undefined) => void'.
src/views/DrefApplicationForm/Operation/index.tsx:343:33 - error TS2322: Type 'string | undefined' is not assignable to type 'number | null | undefined'.
src/views/DrefApplicationForm/Operation/index.tsx:344:33 - error TS2322: Type '(...entries: EntriesAsList<{ title?: string | undefined; id?: number | undefined; status?: 0 | 1 | undefined; national_society?: number | undefined; ifrc?: string | undefined; translation_module_original_language?: string | undefined; ... 133 more ...; anticipatory_actions?: string | undefined; } & Pick<...>>) => void' is not assignable to type '(value: number | undefined, name: "people_per_urban", e?: FormEvent<HTMLInputElement> | undefined) => void'.
src/views/DrefApplicationForm/Operation/index.tsx:350:33 - error TS2322: Type 'string | undefined' is not assignable to type 'number | null | undefined'.
src/views/DrefApplicationForm/Operation/index.tsx:351:33 - error TS2322: Type '(...entries: EntriesAsList<{ title?: string | undefined; id?: number | undefined; status?: 0 | 1 | undefined; national_society?: number | undefined; ifrc?: string | undefined; translation_module_original_language?: string | undefined; ... 133 more ...; anticipatory_actions?: string | undefined; } & Pick<...>>) => void' is not assignable to type '(value: number | undefined, name: "people_per_local", e?: FormEvent<HTMLInputElement> | undefined) => void'.
src/views/Emergencies/Map/index.tsx:201:39 - error TS2352: Conversion of type 'string' to type '{ type: "Point"; coordinates: [number, number]; }' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
src/views/GlobalThreeW/Map/index.tsx:116:27 - error TS2352: Conversion of type 'string' to type '{ type: "Point"; coordinates: [number, number]; }' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
src/views/PerPrioritizationForm/index.tsx:260:21 - error TS2322: Type '{ id: number; area_details: { id: number; title: string; area_num?: number | undefined; }; area: number; component_responses?: { id: number; component: number; rating?: number | null | undefined; ... 4 more ...; climate_environmental_considerations?: string | ... 1 more ... | undefined; }[] | undefined; }[] | undefi...' is not assignable to type '({ id?: number | undefined; area_details?: ({ title?: string | undefined; id?: number | undefined; area_num?: number | undefined; } & Pick<{ id: number; title: string; area_num?: number | undefined; }, never>) | undefined; component_responses?: ({ ...; } & Pick<...>)[] | undefined; } & Pick<...>)[] | undefined'.
src/views/Register/index.tsx:51:41 - error TS2339: Property 'requestBody' does not exist on type '{ parameters: { query: { format?: "csv" | "json" | undefined; }; }; responses: { 200: unknown; }; }'.
src/views/Register/index.tsx:167:22 - error TS2345: Argument of type 'Record<string, unknown> & { [nonFieldError]: string; }' is not assignable to parameter of type 'SetErrorArg<LeafError | ArrayError<any> | ObjectError<any>>'.
src/views/Register/index.tsx:239:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:248:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:258:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:268:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:278:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:291:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:300:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:312:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:325:25 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:335:25 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:345:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:353:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
src/views/Register/index.tsx:361:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
```

**Review by shreeyash07 — APPROVED** [review]

File: src/views/Search/useColumns/index.tsx [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+}
+
+function useColumns(searchResponse: SearchResponse | undefined) {
+    const strings = useTranslation(i18n);
+    const {
+        emergency: emergencyRoute,
+        country: countryRoute,
+    } = useContext(RouteContext);
+
+    const columnMap = {
```
Comment (tnagorra) [inline src/views/Search/useColumns/index.tsx]: Let's use memo

File: src/views/Search/index.tsx [hunk 6]
```diff
@@ -1,211 +1,200 @@
...
 
-        const tableScoreList = mapToList(
-            initialResultMap,
-            (item, key) => ({
-                key: key as ResultKeys,
-                totalItems: item.length,
-                value: Math.max(0, ...(item.map((i) => i.score))),
-            }),
-        );
+            function getNormalizedScore(
```
Comment (tnagorra) [inline src/views/Search/index.tsx]:
```suggestion
            function getAverageScore(
```

File: src/views/Search/ResultTable/index.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                return undefined;
+            }
+
+            if (isNotDefined(maxItems) || mappings.data.length < maxItems) {
+                return mappings.data;
+            }
+
+            const newData = [...mappings.data].splice(0, maxItems);
+
+            return newData;
```
Comment (tnagorra) [inline src/views/Search/ResultTable/index.tsx]:
```suggestion
    return mappings.data.slice(0, maxItems);
```

**Review by tnagorra — CHANGES_REQUESTED** [review]

**Comment by frozenhelium** [comment]
> @frozenhelium @thenav56
> 
> Using server `https://per-ifrc.dev.datafriendlyspace.org/` there are still 25 typeerrors.
> 
> The errors are not just on Register and Dref pages.
> 
> ```
> src/components/ActiveOperationMap/index.tsx:225:39 - error TS2352: Conversion of type 'string' to type '{ type: "Point"; coordinates: [number, number]; }' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
> src/views/DrefApplicationForm/Operation/index.tsx:331:25 - error TS2322: Type 'string | undefined' is not assignable to type 'number | null | undefined'.
> src/views/DrefApplicationForm/Operation/index.tsx:332:25 - error TS2322: Type '(...entries: EntriesAsList<{ title?: string | undefined; id?: number | undefined; status?: 0 | 1 | undefined; national_society?: number | undefined; ifrc?: string | undefined; translation_module_original_language?: string | undefined; ... 133 more ...; anticipatory_actions?: string | undefined; } & Pick<...>>) => void' is not assignable to type '(value: number | undefined, name: "disability_people_per", e?: FormEvent<HTMLInputElement> | undefined) => void'.
> src/views/DrefApplicationForm/Operation/index.tsx:343:33 - error TS2322: Type 'string | undefined' is not assignable to type 'number | null | undefined'.
> src/views/DrefApplicationForm/Operation/index.tsx:344:33 - error TS2322: Type '(...entries: EntriesAsList<{ title?: string | undefined; id?: number | undefined; status?: 0 | 1 | undefined; national_society?: number | undefined; ifrc?: string | undefined; translation_module_original_language?: string | undefined; ... 133 more ...; anticipatory_actions?: string | undefined; } & Pick<...>>) => void' is not assignable to type '(value: number | undefined, name: "people_per_urban", e?: FormEvent<HTMLInputElement> | undefined) => void'.
> src/views/DrefApplicationForm/Operation/index.tsx:350:33 - error TS2322: Type 'string | undefined' is not assignable to type 'number | null | undefined'.
> src/views/DrefApplicationForm/Operation/index.tsx:351:33 - error TS2322: Type '(...entries: EntriesAsList<{ title?: string | undefined; id?: number | undefined; status?: 0 | 1 | undefined; national_society?: number | undefined; ifrc?: string | undefined; translation_module_original_language?: string | undefined; ... 133 more ...; anticipatory_actions?: string | undefined; } & Pick<...>>) => void' is not assignable to type '(value: number | undefined, name: "people_per_local", e?: FormEvent<HTMLInputElement> | undefined) => void'.
> src/views/Emergencies/Map/index.tsx:201:39 - error TS2352: Conversion of type 'string' to type '{ type: "Point"; coordinates: [number, number]; }' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
> src/views/GlobalThreeW/Map/index.tsx:116:27 - error TS2352: Conversion of type 'string' to type '{ type: "Point"; coordinates: [number, number]; }' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
> src/views/PerPrioritizationForm/index.tsx:260:21 - error TS2322: Type '{ id: number; area_details: { id: number; title: string; area_num?: number | undefined; }; area: number; component_responses?: { id: number; component: number; rating?: number | null | undefined; ... 4 more ...; climate_environmental_considerations?: string | ... 1 more ... | undefined; }[] | undefined; }[] | undefi...' is not assignable to type '({ id?: number | undefined; area_details?: ({ title?: string | undefined; id?: number | undefined; area_num?: number | undefined; } & Pick<{ id: number; title: string; area_num?: number | undefined; }, never>) | undefined; component_responses?: ({ ...; } & Pick<...>)[] | undefined; } & Pick<...>)[] | undefined'.
> src/views/Register/index.tsx:51:41 - error TS2339: Property 'requestBody' does not exist on type '{ parameters: { query: { format?: "csv" | "json" | undefined; }; }; responses: { 200: unknown; }; }'.
> src/views/Register/index.tsx:167:22 - error TS2345: Argument of type 'Record<string, unknown> & { [nonFieldError]: string; }' is not assignable to parameter of type 'SetErrorArg<LeafError | ArrayError<any> | ObjectError<any>>'.
> src/views/Register/index.tsx:239:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:248:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:258:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:268:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:278:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:291:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:300:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:312:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:325:25 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:335:25 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:345:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:353:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> src/views/Register/index.tsx:361:21 - error TS2322: Type 'LeafError | ArrayError<any> | ObjectError<any>' is not assignable to type 'ReactNode'.
> ```

Not anymore!

**Review by tnagorra — APPROVED** [review]

**Pull request merged**